man/sbuild.1.in
'\" t
.\" Copyright © 1998       James Troup <james@nocrew.org>
.\" Copyright © 2005-2009  Roger Leigh <rleigh@debian.org>
.\" Copyright © 2008       Simon McVittie <smcv@debian.org>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see
.\" <http://www.gnu.org/licenses/>.
.so defs.man
.TH SBUILD 1 "\*[RELEASE_DATE]" "Version \*[VERSION]" "Debian sbuild"
.SH NAME
sbuild \- build debian packages from source
.SH SYNOPSIS
.B sbuild
.RB [ \-h \[or] \-\-help " \[or] " \-V \[or] \-\-version ]
.RB [ \-v \[or] \-\-verbose " \[or] " \-q \[or] \-\-quiet ]
.RB [ \-D \[or] \-\-debug ]
.RB [ \-A \[or] \-\-arch\-all ]
.RB [ \-\-archive=\fIarchive\fP ]
.RB [ \-d \[or] \-\-dist=\fIdistribution\fP ]
.RB [ \-c \[or] \-\-chroot=\fIchroot\fP ]
.RB [ \-\-chroot\-mode=\fIschroot|sudo|autopkgtest|unshare\fP ]
.RB [ \-\-arch=\fIarchitecture\fP ]
.RB [ \-\-arch\-any " \[or] " \-\-no\-arch\-any ]
.RB [ \-\-build=\fIarchitecture\fP ]
.RB [ \-\-host=\fIarchitecture\fP ]
.RB [ \-\-profiles=\fIprofile[,...]\fP ]
.RB [ \-s \[or] \-\-source ]
.RB [ \-\-force\-orig\-source ]
.RB [ \-\-make\-binNMU=\fIchangelog-entry\fP ]
.RB [ \-\-binNMU=\fINMU-version\fP ]
.RB [ \-\-append\-to\-version=\fIstring\fP ]
.RB [ \-\-binNMU\-timestamp=\fItimestamp\fP ]
.RB [ \-\-binNMU\-changelog=\fIchangelog\fP ]
.RB [ \-\-build\-dir=\fIdirectory\fP ]
.RB [ \-\-add\-depends=\fIdependency\fP ]
.RB [ \-\-add\-conflicts=\fIdependency\fP ]
.RB [ \-\-add\-depends\-arch=\fIdependency\fP ]
.RB [ \-\-add\-conflicts\-arch=\fIdependency\fP ]
.RB [ \-\-add\-depends\-indep=\fIdependency\fP ]
.RB [ \-\-add\-conflicts\-indep=\fIdependency\fP ]
.RB [ \-m \[or] \-\-maintainer=\fImaintainer\fP ]
.RB [ \-e \[or] \-\-uploader=\fIuploader\fP ]
.RB [ \-k \[or] \-\-keyid=\fIkey-id\fP ]
.RB [ \-\-source\-only\-changes ]
.RB [ \-\-no\-source\-only\-changes ]
.RB [ \-j \[or] \-\-jobs=\fIn\fP ]
.RB [ \-\-enable\-network ]
.RB [ \-\-no\-enable\-network ]
.RB [ \-\-debbuildopt=\fIoption\fP ]
.RB [ \-\-debbuildopts=\fIoptions\fP ]
.RB [ \-\-dpkg\-source\-opt=\fIoptions\fP ]
.RB [ \-\-dpkg\-source\-opts=\fIoptions\fP ]
.RB [ \-\-dpkg\-file\-suffix=\fIsuffix\fP ]
.RB [ \-p \[or] \-\-purge=\fPpurge-mode\fP ]
.RB [ \-\-purge\-build=\fPpurge-mode\fP ]
.RB [ \-\-purge\-deps=\fPpurge-mode\fP ]
.RB [ \-\-purge\-session=\fPpurge-mode\fP ]
.RB [ \-b \[or] \-\-batch]
.RB [ \-n \[or] \-\-nolog ]
.RB [ \-\-hide-log-sections=\fPsections\fP ]
.RB [ \-\-clean\-source ]
.RB [ \-\-no\-clean\-source ]
.RB [ \-\-run\-lintian ]
.RB [ \-\-no\-run\-lintian ]
.RB [ \-\-lintian\-opt=\fIoptions\fP ]
.RB [ \-\-lintian\-opts=\fIoptions\fP ]
.RB [ \-\-run\-piuparts ]
.RB [ \-\-no\-run\-piuparts ]
.RB [ \-\-piuparts\-opt=\fIoptions\fP ]
.RB [ \-\-piuparts\-opts=\fIoptions\fP ]
.RB [ \-\-piuparts\-root\-arg=\fIoptions\fP ]
.RB [ \-\-piuparts\-root\-args=\fIoptions\fP ]
.RB [ \-\-run\-autopkgtest ]
.RB [ \-\-no\-run\-autopkgtest ]
.RB [ \-\-autopkgtest\-opt=\fIoptions\fP ]
.RB [ \-\-autopkgtest\-opts=\fIoptions\fP ]
.RB [ \-\-autopkgtest\-root\-arg=\fIoptions\fP ]
.RB [ \-\-autopkgtest\-root\-args=\fIoptions\fP ]
.RB [ \-\-pre\-build\-commands=\fIstring\fP ]
.RB [ \-\-chroot\-setup\-commands=\fIstring\fP ]
.RB [ \-\-chroot\-update\-failed\-commands=\fIstring\fP ]
.RB [ \-\-build\-deps\-failed\-commands=\fIstring\fP ]
.RB [ \-\-starting\-build\-commands=\fIstring\fP ]
.RB [ \-\-finished\-build\-commands=\fIstring\fP ]
.RB [ \-\-build\-failed\-commands=\fIstring\fP ]
.RB [ \-\-chroot\-cleanup\-commands=\fIstring\fP ]
.RB [ \-\-post\-build\-commands=\fIstring\fP ]
.RB [ \-\-post\-build\-failed\-commands=\fIstring\fP ]
.RB [ \-\-anything\-failed\-commands=\fIstring\fP ]
.RB [ \-\-log\-external\-command\-output ]
.RB [ \-\-log\-external\-command\-error ]
.RB [ \-\-setup\-hook=\fIhook-script\fP ]
.RB [ \-\-build\-dep\-resolver=\fIresolver\fP ]
.RB [ \-\-resolve\-alternatives \[or] \-\-no\-resolve\-alternatives ]
.RB [ \-\-extra\-package=\fIpackage.deb|directory\fP ]
.RB [ \-\-extra\-repository=\fIspec\fP ]
.RB [ \-\-extra\-repository\-key=\fIfile.asc\fP ]
.RB [ \-\-build\-path=\fIstring\fP ]
.RB [ \-\-dsc\-dir=\fIstring\fP ]
.RB [ \-\-autopkgtest\-virt\-server=\fIschroot|lxc|chroot|qemu|ssh|lxd|podman|docker|incus|unshare\fP ]
.RB [ \-\-autopkgtest\-virt\-server\-opt=\fIstring\fP ]
.RB [ \-\-autopkgtest\-virt\-server\-opts=\fIoptions\fP ]
.RB [ \-\-purge\-extra\-packages\fP ]
.RB [ \-\-bd\-uninstallable\-explainer=\fIdose3|apt|none\fP ]
.RB [ PACKAGE [ .dsc ]]
.SH DESCRIPTION
\fBsbuild\fR rebuilds Debian binary packages from the corresponding Debian
source, installing any missing source dependencies.  The build takes place in a
dedicated clean build environment, rather than on the host system. For an
overview of the supported chroot backends see the section \fBCHROOT MODES\fR.
.PP
\fBsbuild\fR can fetch the Debian source over a network, or it can use
locally available sources.
.PP
sbuild is given a packages to process as the argument \fBPACKAGE[.dsc]\fR.
This argument is in the form of either a debianized package source directory, a
source package name along with a version in the form \fIpackage_version\fP, a
source package name, or a .dsc file. If no arguments are given, the current
working directory is passed as an argument.
.PP
For arguments given as source directories, dpkg\-source is first run to produce
a source .dsc file. Then, the package is built using the .dsc produced. For
arguments in the form \fIpackage_version\fP or \fIpackage\fP, apt is used to
download the source package. For arguments given as a .dsc file, sbuild builds
the source packages directly. For .dsc files in remote locations, the source
packages are downloaded first, then built.
.PP
It is also possible to run external commands with sbuild. See the section
\fBEXTERNAL COMMANDS\fR for more on this.
.PP
\fBsbuild\fR mails the build logs to a user.  It is configured by the
configuration files \fI/etc/sbuild/sbuild.conf\fP and \fI$XDG_CONFIG_HOME/sbuild/config.pl\fP
(preferred over \fI~/.sbuildrc\fP). If \fI$XDG_CONFIG_HOME\fP is either not set
or empty, \fI~/.config/sbuild/config.pl\fP will be used. An example sbuildrc is available in
\fI/usr/share/doc/sbuild/examples/example.sbuildrc\fP.
A custom path to a configuration file can also be specified through setting the
\fBSBUILD_CONFIG\fP environment variable to the path of an additional
configuration file.
.PP
You can build either using a local package with its .dsc file or a
remote one by specifying an explicit dpkg version.
.SH OPTIONS
Options set on the command line overwrite settings made in the configuration
file.
.TP
.BR \-h ", " \-\-help
Display this manual.
.TP
.BR \-V ", " \-\-version
Print version information.
.TP
.BR "\-\-add\-depends=\fIdependency\fP"
.TP
.BR "\-\-add\-conflicts=\fIdependency\fP"
.TP
.BR "\-\-add\-depends\-arch=\fIdependency\fP"
.TP
.BR "\-\-add\-conflicts\-arch=\fIdependency\fP"
.TP
.BR "\-\-add\-depends\-indep=\fIdependency\fP"
.TP
.BR "\-\-add\-conflicts\-indep=\fIdependency\fP"
These options add a build dependencies to the source package being built, in
addition to the build dependency information specified in debian/control.
These dependencies will be concatenated directly to the Build\-Depends,
Build\-Conflicts, Build\-Depends\-Arch, Build\-Conflicts\-Arch, Build\-Depends\-Indep
and Build\-Conflicts\-Indep dependencies, respectively.  The options may be used
any number of times to add multiple dependencies.  The format is identical to
the format used in debian/control.
These command line options append to the \fBMANUAL_DEPENDS\fP, \fBMANUAL_CONFLICTS\fP, \fBMANUAL_DEPENDS_ARCH\fP, \fBMANUAL_CONFLICTS_ARCH\fP, \fBMANUAL_DEPENDS_INDEP\fP and \fBMANUAL_CONFLICTS_INDEP\fP configuration variables, respectively. See
.BR sbuild.conf (5)
for more information.
.TP
.BR "\-\-arch=\fIarchitecture\fP"
Build using the architecture specified.  A chroot named
\fI$distribution\-$arch\-sbuild\fP or \fI$distribution\-arch\fP is searched for,
in that order of preference.  The chroot must be installed and configured
appropriately to build as that architecture, e.g. using
\fIpersonality=linux32\fP to build i386 packages on an amd64 system.  Note that
this option is equivalent to "\-\-host=architecture \-\-build=architecture".
If you specify an architecture that your CPU cannot execute, make sure that
you have qemu-user with binfmt support installed to emulate the foreign
architecture. Since build and host architecture are equivalent, this option
will result in a native compilation, even if the native compilation is being
emulated transparently using qemu.
This command line option sets the \fBHOST_ARCH\fP and \fBBUILD_ARCH\fP configuration variables. See
.BR sbuild.conf (5)
for more information.
.TP
.BR "\-\-host=\fIarchitecture\fP"
Build using the host architecture specified. Defaults to the native architecture.
The host architecture is the architecture that sbuild will build packages for.
If the specified architecture is different from the build architecture, then
a cross-build is being performed.
If $host and $build don't match, a
chroot named \fI$distribution\-$build\-$host\-sbuild\fP or \fI$distribution\-$build\-$host\fP
is searched for, falling back to \fI$distribution\-$build\-sbuild\fP or
\fI$distribution\-$build\fP, in that order of preference.
This command line option sets the \fBHOST_ARCH\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR "\-\-build=\fIarchitecture\fP"
Build using the build architecture specified. Defaults to the native architecture.
The build architecture is the architecture of the chroot that the package will
be built in. If you want to perform a native build, set the host architecture
to the same value or use the --arch option.
This command line option sets the \fBBUILD_ARCH\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR \-A ", " "\-\-arch\-all"
Also build Architecture: all packages. This is the default behaviour for native builds. This option is the opposite of
\-\-no\-arch\-all.
See the section
.BR "BUILD ARTIFACTS"
for more information.
This command line option sets the \fBBUILD_ARCH_ALL\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR "\-\-no\-arch\-all"
Do not build Architecture: all packages. This is the default behaviour for cross builds. This
option is the opposite of \-\-arch\-all.
See the section
.BR "BUILD ARTIFACTS"
for more information.
This command line option sets the \fBBUILD_ARCH_ALL\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR "\-\-arch\-any"
Build Architecture: any packages. This is the default behavior. This option is
the opposite of \-\-no\-arch\-any.
See the section
.BR "BUILD ARTIFACTS"
for more information.
This command line option sets the \fBBUILD_ARCH_ANY\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR "\-\-no\-arch\-any"
Do not build Architecture: any packages. This option is the opposite
of \-\-arch\-any and only useful when used together with \-\-arch\-all
or \-\-source.
See the section
.BR "BUILD ARTIFACTS"
for more information.
This command line option sets the \fBBUILD_ARCH_ANY\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR \-b ", " "\-\-batch"
Operate in batchmode, i.e. write a build-progress file during execution
and files on shutdown to facilitate a clean restart.
This command line option sets the \fBBATCH_MODE\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR \-c ", " "\-\-chroot=\fIchroot\fP"
Specifies the chroot to use. The effect of this option depends on the selected
chroot mode.  With the \fBschroot\fP chroot mode, this option specifies the
schroot name or alias to use. If not specified, the default is the first of
schroot name or alias that matches \fI$distribution\-$arch\-sbuild\fP,
\fI$distribution\-sbuild\fP, \fI$distribution\-$arch\fP or \fI$distribution\fP
that exists.  With the \fBsudo\fP chroot mode, this option specifies the chroot
directory to use.  The directory is either expected in /etc/sbuild/chroot (in
buildd sbuild mode) or in the build directory (see \-\-build\-dir), prefixed with "chroot\-" (in
user sbuild mode, the default). If not specified, the default is to search for
a directory in the respective locations named in the same way as for the
schroot mode.  With the \fBunshare\fP chroot mode, if this option is a path,
then it specifies the location of the chroot tarball directly. Otherwise, a
tarball with equal basename from ~/.cache/sbuild will be used. If not
specified, the default is to search for a tarball named in the same way as for
the schroot mode under ~/.cache/sbuild.  With the \fBautopkgtest\fP chroot mode
this option has no effect.  The \-\-autopkgtest\-virt\-server\-opts are used to pick
the chroot in autopkgtest chroot mode.
This command line option sets the \fBCHROOT\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR "\-\-chroot\-mode=\fIschroot|sudo|autopkgtest|unshare\fP"
Select the desired chroot mode. Four values are possible: schroot (the
default), sudo (which uses sudo to execute chroot in a directory from
/etc/sbuild/chroot or ./chroot), autopkgtest which uses the autopkgtest\-virt\-* binaries
(selectable via the \-\-autopkgtest\-virt\-server option) and unshare (which uses linux
namespaces for chroot and doesn't require superuser privileges).
See the section
.BR "CHROOT MODES"
for more information.
This command line option sets the \fBCHROOT_MODE\fP configuration variable. See
.BR sbuild.conf (5)
for more information.
.TP
.BR \-d ", " "\-\-dist=\fIdistribution\fP"
Explicitly set the distribution for the package build. This will be selecting
the correct chroot to use and also sets the value of the Distribution field in
the created .changes file. Setting this option is necessary when giving sbuild
a .dsc file or a plain source package name to build. In the latter case it

debian/NEWS
sbuild (0.87.1) unstable; urgency=medium

    Default build path is /build/reproducible-path and switch off log
    filtering. To restore the previous behaviour, use:

        $build_path = '';
        $log_filter = 1;

    Changing the default build path aligns default sbuild behaviour with sbuild
    as run on Debian buildds and thus uploaded .buildinfo files can now
    exactly mimic those that will be created by the buildds.

    This change will makes it impossible to do concurrent builds of the same
    package with the same version using the schroot backend. If you want to
    do multiple builds of the same package, consider to either:

     - vary the version for example with --make-binNMU=...
     - manually set the build path via ~/.sbuildrc or --build-path
     - don't use the schroot backend

    With non-randomized build paths, log filtering does not serve its original
    purpose anymore (easy diffing of two build logs). Disabling log filtering
    will also improve readability of the build logs for newcomers who are not
    familiar with the placeholders.

 -- Johannes Schauer Marin Rodrigues <josch@debian.org>  Wed, 13 Nov 2024 08:00:24 +0100

sbuild (0.87.0) unstable; urgency=medium

    This release introduces an experimental new feature for the unshare
    backend which is supposed to make sbuild more easy to use because
    sbuild will manage the chroot tarballs itself.
    With unshare-mode, sbuild will now automatically create a chroot tarball
    upon startup using mmdebstrap if it cannot find one in ~/.cache/sbuild.
    This feature can be disabled by setting $unshare_mmdebstrap_auto_create=0;
    in your ~/.sbuildrc. If you like the feature and you would like to keep
    the tarballs sbuild creates so that they can be re-used in subsequent
    invocations of sbuild, you can set $unshare_mmdebstrap_keep_tarball=1; in
    your ~/.sbuildrc. This will also make sbuild update an existing tarball (by
    re-creating it) if it has become too old. The maximum age of a chroot
    tarball can be specified by setting $unshare_mmdebstrap_max_age to a
    maximum age value in seconds (default is 604800, which is one week).
    To customize the chroot tarball creation, refer to the setting
    UNSHARE_MMDEBSTRAP_EXTRA_ARGS in your sbuild.conf(5).

 -- Johannes Schauer Marin Rodrigues <josch@debian.org>  Thu, 24 Oct 2024 16:39:50 +0200

sbuild (0.75.0-1) unstable; urgency=medium

    sbuild now builds arch:all packages by default.
    sbuild now runs lintian by default.

 -- Michael Stapelberg <stapelberg@debian.org>  Wed, 21 Mar 2018 22:25:31 +0100


sbuild (0.73.0-3) unstable; urgency=medium

    Generating a public/private keypair using "sbuild-update --keygen" for
    signing the sbuild internal dummy repository is now optional. You are only
    required to generate the keys if you plan to build packages for Debian
    Squeeze or older.

    If you only build packages for Debian Wheezy or newer, then
    /var/lib/sbuild/apt-keys/sbuild-key.sec and
    /var/lib/sbuild/apt-keys/sbuild-key.pub can be safely removed from
    your system.

 -- Johannes Schauer <josch@debian.org>  Tue, 24 Jan 2017 10:51:16 +0100

sbuild (0.69.0-1) unstable; urgency=medium

    With this release, logically opposite boolean command line arguments cannot
    be specified together anymore. Attempting to do so will result in an error.

    The remaining changes in this version only concern users of sbuild external
    command hooks involving the %SBUILD_CHROOT_DIR percent escape in
    --pre-build-commands or --post-build-commands.

    With this release, the %SBUILD_CHROOT_DIR percent escape is now deprecated.
    It can still be used but it will only work for sbuild backends that work on
    the same file system as the sbuild process. With the addition of more
    backends like the autopkgtest backend sbuild is now able to build inside
    qemu virtual machines, containers and on remote hosts via ssh. In these
    cases the %SBUILD_CHROOT_DIR variable doesn't make sense anymore because
    the sbuild process cannot directly access the build environment anymore.

    Instead, the %SBUILD_CHROOT_EXEC percent escape can be used to run an
    arbitrary command inside the build chroot independent of the used backend.
    See the sbuild manual page of how to utilize %SBUILD_CHROOT_EXEC to even
    transfer files between the host running sbuild and the chroot backend.

 -- Johannes Schauer <josch@debian.org>  Mon, 23 Jan 2017 19:54:26 +0100

sbuild (0.58.3-1) unstable; urgency=low

    The sbuild group was previously created as a non-sytem group.
    When upgrading from older versions of sbuild, the group will be
    deleted and re-created as a system group.

    Group membership will be transferred to the new group, and system
    files and directories owned by group sbuild will be updated to be
    owned by the new group.  Files and directories which are set to be
    owned by group sbuild on the system or inside any build chroots
    will need to be manually updated.  The group database inside any
    build chroots may also need updating.  If using schroot to access
    chroots, the group database should normally be updated
    automatically, as should file and directory ownership.

 -- Roger Leigh <rleigh@debian.org>  Sun, 17 May 2009 19:16:30 +0100

sbuild (0.58.0) unstable; urgency=low

    The configuration option @no_auto_upgrade, which was no longer
    used, has been removed.  To prevent packages being upgraded
    automatically, please use sbuild-hold.  See sbuild-hold(1).

    The configuration option $chroot_mode has been restored, allowing
    chroot access by both schroot (which remains the default) and
    sudo.

    A new configuration option $chroot_split has been added.  This
    allows both schroot and sudo chroot access to be modified to run
    apt/dpkg on the host system.  Note that schroot needs sudo to run
    commands as root on the host system.


lib/Sbuild.pm
#
# Sbuild.pm: library for sbuild
# Copyright © 2005 Ryan Murray <rmurray@debian.org>
# Copyright © 2005-2008 Roger Leigh <rleigh@debian.org
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <http://www.gnu.org/licenses/>.
#
#######################################################################

package Sbuild;

use Sbuild::Sysconfig;

use strict;
use warnings;
use POSIX;
use FileHandle;
use Filesys::Df qw();
use Time::Local;
use IO::Zlib;
use MIME::Base64;
use Dpkg::Control;
use Dpkg::Checksums;
use POSIX qw(locale_h);

BEGIN {
	use Exporter ();
	our (@ISA, @EXPORT);

	@ISA = qw(Exporter);

	@EXPORT = qw($debug_level $devnull binNMU_version parse_date isin
	  copy dump_file check_packages help_text version_text
	  usage_error send_mail debug debug2 df check_unshare
	  check_group_membership dsc_files dsc_pkgver shellescape strftime_c);
}

our $devnull;
our $debug_level = 0;

BEGIN {
	# A file representing /dev/null
	if (!open($devnull, '+<', '/dev/null')) {
		die "Cannot open /dev/null: $!\n";
	}
}

sub binNMU_version ($$$);
sub parse_date ($);
sub isin ($@);
sub copy ($);
sub dump_file ($);
sub check_packages ($$);
sub help_text ($$);
sub version_text ($);
sub usage_error ($$);
sub debug (@);
sub debug2 (@);
sub check_unshare($$);
sub check_group_membership();
sub dsc_files ($);
sub shellescape ($);
sub strftime_c ($@);

sub binNMU_version ($$$) {
	my $v                 = shift;
	my $binNMUver         = shift;
	my $append_to_version = shift;

	my $ver = $v;
	if (defined($append_to_version) && $append_to_version) {
		$ver .= $append_to_version;
	}
	if (defined($binNMUver) && $binNMUver) {
		$ver .= "+b$binNMUver";
	}
	return $ver;
}

my %monname = (
	'jan', 0, 'feb', 1, 'mar', 2, 'apr', 3, 'may', 4,  'jun', 5,
	'jul', 6, 'aug', 7, 'sep', 8, 'oct', 9, 'nov', 10, 'dec', 11
);

sub parse_date ($) {
	my $text = shift;

	return 0 if !$text;
	die "Cannot parse date: $text\n"
	  if $text !~ /^(\d{4}) (\w{3}) (\d+) (\d{2}):(\d{2}):(\d{2})$/;
	my ($year, $mon, $day, $hour, $min, $sec) = ($1, $2, $3, $4, $5, $6);
	$mon =~ y/A-Z/a-z/;
	die "Invalid month name $mon" if !exists $monname{$mon};
	$mon = $monname{$mon};
	return timegm($sec, $min, $hour, $day, $mon, $year);
}

sub isin ($@) {
	my $val = shift;
	return grep($_ eq $val, @_);
}

sub copy ($) {
	my $r = shift;
	my $new;

	if (ref($r) eq "HASH") {
		$new = {};
		foreach (keys %$r) {
			$new->{$_} = copy($r->{$_});
		}
	} elsif (ref($r) eq "ARRAY") {
		my $i;
		$new = [];
		for ($i = 0 ; $i < @$r ; ++$i) {
			$new->[$i] = copy($r->[$i]);
		}
	} elsif (!ref($r)) {
		$new = $r;
	} else {
		die "unknown ref type in copy\n";
	}

	return $new;
}

sub dump_file ($) {
	my $file = shift;

	if (-r "$file"
		&& open(SOURCES, "<$file")) {

		print
"   +------------------------------------------------------------------------\n";
		while (<SOURCES>) {
			chomp;
			print "   |$_\n";
		}
		print
"   +------------------------------------------------------------------------\n";
		close(SOURCES) or print "Failed to close $file\n";
	} else {
		print "W: Failed to open $file\n";
	}
}

# set and list saved package list (used by sbuild-checkpackages)
sub check_packages ($$) {
	my $session = shift;
	my $mode    = shift;

	my $package_checklist = $session->get_conf('PACKAGE_CHECKLIST');
	my $chroot_dir        = $session->get('Location');

	my (@status, @ref, @install, @remove);

	my $pipe = $session->pipe_command({
			COMMAND => [
				'dpkg-query', '--show',
				'--showformat=${Package} ${db:Status-Status}\n'
			] });
	while (<$pipe>) {
		chomp;
		my @token   = split / /, $_;
		my $pkgname = shift @token;
		my $state   = shift @token;
		next if $state ne "installed";
		push @status, $pkgname;
	}
	if (!close $pipe) {
		print STDERR "Error reading dpkg status file in chroot: $!\n";
		return 1;
	}
	@status = sort @status;
	if (!@status) {
		print STDERR "dpkg status file is empty\n";
		return 1;
	}

	if ($mode eq "set") {
		if (!open WREF, "> $chroot_dir/$package_checklist") {
			print STDERR
"Can't write reference status file $chroot_dir/$package_checklist: $!\n";
			return 1;
		}
		foreach (@status) {
			print WREF "$_\n";
		}
		if (!close WREF) {
			print STDERR "Error writing reference status file: $!\n";
			return 1;
		}
	} else {    # "list"
		if (!open REF, "< $chroot_dir/$package_checklist") {
			print STDERR
"Can't read reference status file $chroot_dir/$package_checklist: $!\n";
			return 1;
		}
		while (<REF>) {
			chomp;
			push @ref, $_;
		}
		if (!close REF) {
			print STDERR "Error reading reference status file: $!\n";
			return 1;
		}

		@ref = sort @ref;
		if (!@ref) {
			print STDERR "Reference status file is empty\n";
			return 1;
		}

		print "DELETE             ADD\n";
		print "--------------------------------------\n";
		my $i = 0;
		my $j = 0;

		while ($i < scalar @status && $j < scalar @ref) {

			my $c = $status[$i] cmp $ref[$j];
			if ($c < 0) {
				# In status, not reference; remove.
				print "$status[$i]\n";
				$i++;
			} elsif ($c > 0) {
				# In reference, not status; install.
				print "                   $ref[$j]\n";
				$j++;
			} else {
				# Identical; skip.
				$i++;
				$j++;
			}
		}

		# Print any remaining elements
		while ($i < scalar @status) {
			print "$status[$i]\n";
			$i++;
		}
		while ($j < scalar @ref) {
			print "                   $ref[$j]\n";
			$j++;
		}
	}
}

sub help_text ($$) {
	my $section = shift;
	my $page    = shift;

	system('man', '--', $section, $page);
	exit 0;
}

sub version_text ($) {
	my $program = shift;

	print <<"EOF";
$program (Debian sbuild) $Sbuild::Sysconfig::version ($Sbuild::Sysconfig::release_date)

Written by Roman Hodek, James Troup, Ben Collins, Ryan Murray, Rick
Younie, Francesco Paolo Lovergine, Michael Banck, Roger Leigh and
Andres Mejia.

Copyright © 1998-2000 Roman Hodek <roman\@hodek.net>
          © 1998-1999 James Troup <troup\@debian.org>
	  © 2003-2006 Ryan Murray <rmurray\@debian.org>
	  © 2001-2003 Rick Younie <younie\@debian.org>
	  © 2003-2004 Francesco Paolo Lovergine <frankie\@debian.org>
	  © 2005      Michael Banck <mbanck\@debian.org>
	  © 2005-2010 Roger Leigh <rleigh\@debian.org>
	  © 2009-2010 Andres Mejia <mcitadel\@gmail.com>

This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
EOF
	exit 0;
}

# Print an error message about incorrect command-line options
sub usage_error ($$) {
	my $program = shift;
	my $message = shift;

	print STDERR "E: $message\n";
	print STDERR
"I: Run '$program --help' to list usage example and all available options\n";
	exit 1;
}

sub send_mail ($$$$) {
	my $conf    = shift;
	my $to      = shift;
	my $subject = shift;
	my $file    = shift;
	local (*MAIL, *F);

	if (!open(F, "<$file")) {
		warn "Cannot open $file for mailing: $!\n";
		return 0;
	}
	local $SIG{'PIPE'} = 'IGNORE';

	if (!open(MAIL, "|" . $conf->get('MAILPROG') . " -oem $to")) {
		warn "Could not open pipe to " . $conf->get('MAILPROG') . ": $!\n";
		close(F);
		return 0;
	}

	print MAIL "From: " . $conf->get('MAILFROM') . "\n";
	print MAIL "To: $to\n";
	print MAIL "Subject: $subject\n";
	print MAIL "Content-Type: text/plain; charset=UTF-8\n";
	print MAIL "Content-Transfer-Encoding: 8bit\n";
	print MAIL "\n";
	while (<F>) {
		print MAIL "." if $_ eq ".\n";
		print MAIL $_;
	}

	close(F);
	if (!close(MAIL)) {
		warn $conf->get('MAILPROG') . " failed (exit status $?)\n";
		return 0;
	}
	return 1;
}

# Note: split to stderr
sub debug (@) {

	# TODO: Add debug level checking.
	if ($debug_level) {
		print STDERR "D: ", @_;
	}
}

sub debug2 (@) {

	# TODO: Add debug level checking.
	if ($debug_level && $debug_level >= 2) {
		print STDERR "D2: ", @_;
	}
}

sub df {
	my $dir = shift;

	my $stat = Filesys::Df::df($dir);

	return $stat->{bfree} if (defined($stat));

	# This only happens if $dir was not a valid file or directory.
	return 0;
}

sub check_unshare($$) {
	my $prog = shift;
	my $conf = shift;

	if ($conf->get('CHROOT_MODE') eq 'unshare') {
		print
		  "$prog is only useful for chroot_mode=\"schroot\" and \"sudo\".\n";
		print "\n";
		print "With the \"unshare\" mode, it is without function.\n";
		print "Instead, recreate tarballs from scratch or let sbuild \n";
		print
"manage them for you and use mmdebstrap_extra_args to do the custom setup.\n";
		exit 1;
	}
}

sub check_group_membership () {
	# Skip for root
	return if ($< == 0);

	my $user = getpwuid($<);
	my ($name, $passwd, $gid, $members) = getgrnam("sbuild");

	if (!$gid) {
		die "Group sbuild does not exist";
	}

	my $in_group = 0;
	my @groups   = getgroups();
	push @groups, getgid();
	foreach (@groups) {
		($name, $passwd, $gid, $members) = getgrgid($_);
		$in_group = 1 if defined($name) && $name eq 'sbuild';
	}

	if (!$in_group) {
		print STDERR
"User $user is not currently an effective member of group sbuild. Please run:\n";
		print STDERR " sudo sbuild-adduser $user\n";
		print STDERR
"And then either log out and log in again or use `newgrp sbuild` to gain sbuild group privileges\n";
		exit(1);
	}

	return;
}

sub dsc_files ($) {
	my $dsc = shift;

	debug("Parsing $dsc\n");
	my $pdsc = Dpkg::Control->new(type => CTRL_PKG_SRC);
	$pdsc->set_options(allow_pgp => 1);
	if (!$pdsc->load($dsc)) {
		print STDERR "Could not parse $dsc\n";
		return undef;
	}

	my $csums = Dpkg::Checksums->new();
	$csums->add_from_control($pdsc, use_files_for_md5 => 1);
	return $csums->get_files();
}

sub dsc_pkgver ($) {
	my $dsc = shift;

	debug("Parsing $dsc\n");
	my $pdsc = Dpkg::Control->new(type => CTRL_PKG_SRC);
	$pdsc->set_options(allow_pgp => 1);
	if (!$pdsc->load($dsc)) {
		print STDERR "Could not parse $dsc\n";
		return undef;
	}

	return ($pdsc->{'Source'}, $pdsc->{'Version'});
}

# avoid dependency on String::ShellQuote by implementing the mechanism
# from python's shlex.quote function
sub shellescape ($) {
	my $string = shift;
	if (length $string == 0) {
		return "''";
	}
	# search for occurrences of characters that are not safe
	# the 'a' regex modifier makes sure that \w only matches ASCII
	if ($string !~ m/[^\w@\%+=:,.\/-]/a) {
		return $string;
	}
	# wrap the string in single quotes and handle existing single quotes by
	# putting them outside of the single-quoted string
	$string =~ s/'/'"'"'/g;
	return "'$string'";
}

# this function uses strftime to format a timestamp as a string but makes sure
# to use the C locale to do so instead of the system locale
sub strftime_c ($@) {
	my $format = shift;
	my @time   = @_;

	my $old_locale = setlocale(LC_TIME);
	setlocale(LC_TIME, "C.UTF-8");
	my $ret = strftime $format, @time;
	setlocale(LC_TIME, $old_locale);

	return $ret;
}

1;
